internal/controller/reinstall/reinstall.go

```diff
@@ -0,0 +1,378 @@
+		zap.String("ClusterInstance", client.ObjectKeyFromObject(clusterInstance).String()),
+		zap.String("ResourceVersion", clusterInstance.ResourceVersion),
+		zap.String("ReinstallGeneration", reinstallSpec.Generation),
+		zap.String("PreservationMode", string(reinstallSpec.PreservationMode)),
+	)
+
+	// Nothing to do if the reinstall request is processed
+	if isRequestProcessed(clusterInstance) {
+		return reconcile.Result{}, nil
+	}
+
+	reinstallRequestProcessedCondition := findReinstallStatusCondition(clusterInstance,
+		v1alpha1.ReinstallRequestProcessed)
+
+	if reinstallRequestProcessedCondition == nil || isNewRequest(clusterInstance) {
+		log.Info("New reinstall request detected")
+		if err = initializeReinstallStatus(ctx, r.Client, log, clusterInstance); err != nil {
+			log.Error("Failed to initialize Reinstall Status", zap.Error(err))
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	log.Info("Processing reinstall request")
+
+	// Ensure that ReinstallRequestProcessed InProgress is set after initialization
+	if reinstallRequestProcessedCondition.Reason == string(v1alpha1.Initialized) {
+		patch := client.MergeFrom(clusterInstance.DeepCopy())
+		clusterInstance.Status.Reinstall.RequestStartTime = metav1.Now()
+		reinstallRequestProcessedCondition = reinstallRequestProcessedConditionStatus(
+			metav1.ConditionFalse, v1alpha1.InProgress, fmt.Sprintf("Processing reinstall request for generation %s",
+				reinstallSpec.Generation))
+
+		setReinstallStatusCondition(clusterInstance, *reinstallRequestProcessedCondition)
+
+		if err = patchClusterInstanceReinstallStatus(ctx, r.Client, log, clusterInstance, patch); err != nil {
+			log.Sugar().Errorf("Failed to update reinstall condition '%s'", reinstallRequestProcessedCondition.Type,
+				zap.Error(err))
+			return reconcile.Result{}, err
+		}
+		log.Sugar().Infof("Updated reinstall condition '%s' successfully", reinstallRequestProcessedCondition.Type)
+
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	result = reconcile.Result{}
+	err = nil
+	updateEndtime := false
+
+	defer func() {
+		patch := client.MergeFrom(clusterInstance.DeepCopy())
+		if updateEndtime {
+			clusterInstance.Status.Reinstall.RequestEndTime = metav1.Now()
+			// Add hold annotation for manual intervention (CNF-15719)
+		}
+
+		changed := setReinstallStatusCondition(clusterInstance, *reinstallRequestProcessedCondition)
+		if changed || updateEndtime {
+			updateErr := patchClusterInstanceReinstallStatus(ctx, r.Client, log, clusterInstance, patch)
+			if updateErr != nil {
+				if err == nil {
+					err = updateErr
+				}
+			}
+		}
+	}()
+
+	tasks := []struct {
+		Description string
+		Run         func(context.Context, *zap.Logger, *v1alpha1.ClusterInstance) (reconcile.Result, error)
+	}{
+		{"Validating reinstall request", r.ensureValidReinstallRequest},
+		{"Deleting rendered manifests", r.ensureRenderedManifestsAreDeleted},
+	}
+	for _, task := range tasks {
```

Comment:
I feel like this is a bit unnecessary as a way to deduplicate error handling.
Would it not be clearer to put the error handling in a function and directly call the `ensure*` functions separately?

Reply:
We can do that, however, there will be quite a lot of duplicated code (with a very long `ProcessRequest` function) once I add the preservation & restoration bits (3 additional tasks) in the next PR. 

Furthermore, I feel the way I have it here makes it more readable (less verbose) and easier to follow.

Reply:
That's fair. I feel like this might be justified if it were more than two, but that's hard to see without all the relevant code :wink: